File: apps/orderbook-engine/orderbook.ts
import { type_Orderbook, type_Price, type_AskSide, type_BidSide, type_Quantity, type_Depth } from "./types"

export class Orderbook {

    bids: type_BidSide;
    asks: type_AskSide;
    currentPrice: type_Price;
    market: string

    static bidComparator(a: number[], b: number[]): number {
        return b[0] - a[0];
    }

    static askComparator(a: number[], b: number[]): number {
        return a[0] - b[0];
    }

    constructor(bids: type_BidSide, asks: type_AskSide, market: string) {
        this.asks = asks.sort(Orderbook.askComparator);
        this.market = market;
        this.bids = bids.sort(Orderbook.bidComparator);
        this.currentPrice = 0;
    }

    public printDepth(): void {
        console.log("bids                 asks")
        for (let i = 0; i < Math.max(this.bids.length, this.asks.length); i++) {
            const bid = this.bids[i] ? `${this.bids[i][0]} (${this.bids[i][1]})` : "      ";
            const ask = this.asks[i] ? `${this.asks[i][0]} (${this.asks[i][1]})` : "      ";
            console.log(`${bid}   |   ${ask}`);
        }
        console.log(this.currentPrice)
    }

    public getDepth(): type_Depth {
        const depth: any = {
            bids: [],
            asks: [],
            current_price: Number
        };

        for (let i = 0; i < Math.max(this.bids.length, this.asks.length); i++) {
            const bid = this.bids[i] ? { price: this.bids[i][0], quantity: this.bids[i][1] } : 0;
            const ask = this.asks[i] ? { price: this.asks[i][0], quantity: this.asks[i][1] } : 0;

            if (bid) depth.bids.push(bid);
            if (ask) depth.asks.push(ask);
            depth.current_price = this.currentPrice;
        }

        return depth;
    }

    public buy(price: type_Price, quantity: type_Quantity): void {
        if (!this.bids.length) { this.bids.push([price, quantity]); return; };

        const bestAsk: type_Price[] = this.asks[0];
        const bestBid: type_Price[] = this.bids[0];

        if (price >= bestAsk[0]) {
            let tempQuantity = quantity;

            for (const value of this.asks) {
                if (tempQuantity >= value[1]) {
                    tempQuantity -= value[1];
                    this.asks = this.asks.filter(num => num[0] !== value[0]);
                    this.currentPrice = value[0];
                } else if (tempQuantity < value[1] && price >= value[0]) {
                    value[1] -= tempQuantity;
                    this.currentPrice = value[0];
                    tempQuantity = 0;
                    break;
                }
            }

            if (tempQuantity > 0) {
                this.bids.push([price, tempQuantity]);
                this.bids.sort(Orderbook.bidComparator);
            }

        } else if (price < bestAsk[0]) {
            let flag = false;

            if (price > bestBid[0]) {
                this.bids.push([price, quantity]);
                this.bids.sort(Orderbook.bidComparator);
            } else if (price <= bestBid[0]) {
                for (const value of this.bids) {
                    if (price == value[0]) {
                        flag = true;
                        value[1] += quantity;
                        break;
                    }
                }

                if (!flag) {
                    this.bids.push([price, quantity]);
                    this.bids.sort(Orderbook.bidComparator);
                }
            }
        }

        if (this.currentPrice === 0) {
            this.currentPrice = bestAsk ? bestAsk[0] : this.currentPrice;
        }
    }

    public sell(price: type_Price, quantity: type_Quantity): void {
        if (!this.asks.length) { this.asks.push([price, quantity]); return; };
        const bestAsk: type_Price[] = this.asks[0];
        const bestBid: type_Price[] = this.bids[0];

        if (price <= bestBid[0]) {
            let tempQuantity = quantity;

            for (const value of this.bids) {
                if (tempQuantity >= value[1]) {
                    tempQuantity -= value[1];
                    this.bids = this.bids.filter(num => num[0] !== value[0]);
                    this.currentPrice = value[0];
                } else if (tempQuantity < value[1] && price <= value[0]) {
                    value[1] -= tempQuantity;
                    this.currentPrice = value[0];
                    tempQuantity = 0;
                    break;
                }
            }

            if (tempQuantity > 0) {
                this.asks.push([price, tempQuantity]);
                this.asks.sort(Orderbook.askComparator);
            }

        } else if (price > bestBid[0]) {
            let flag = false;

            if (price < bestAsk[0]) {
                this.asks.push([price, quantity]);
                this.asks.sort(Orderbook.askComparator);
            } else if (price >= bestAsk[0]) {
                for (const value of this.asks) {
                    if (price == value[0]) {
                        flag = true;
                        value[1] += quantity;
                        this.currentPrice = value[0];
                        break;
                    }
                }

                if (!flag) {
                    this.asks.push([price, quantity]);
                    this.asks.sort(Orderbook.askComparator);
                }
            }
        }

        if (this.currentPrice === 0) {
            this.currentPrice = bestBid ? bestBid[0] : this.currentPrice;
        }
    }
}

const bids = [[144.24, 0.80], [144.26, 116.81], [144.25, 69.87], [144.23, 138.67], [144.22, 34.65], [144.21, 0.03], [144.20, 138.68], [144.19, 34.73], [144.18, 138.71]];
const asks = [[144.35, 0.21], [144.37, 138.73], [144.38, 148.96], [144.34, 34.08], [144.33, 34.67], [144.31, 138.66], [144.30, 34.65], [144.29, 17.32], [144.28, 85.55]];

export const sol_usdc_ob: type_Orderbook = new Orderbook(bids, asks, 'SOL/USDC');
export const eth_usdc_ob: type_Orderbook = new Orderbook([], [], "ETH/USDC");

File: 2_orderbook-engine/orderbook/orderbook.go
package orderbook

import (
	"container/heap"
	// "debug/elf"
)

// some basic initialisations for orderbook
type Order struct {
	Kind     string
	Market   string
	Price    float64
	Quantity int
}

type OrderEntry struct {
	Price    float64
	Quantity float64
}

type MaxHeap []*OrderEntry

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i].Price > h[j].Price }

func (h MaxHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x interface{}) { *h = append(*h, x.(*OrderEntry)) }
func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

type MinHeap []*OrderEntry

func (h MinHeap) Len() int           { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i].Price < h[j].Price }

func (h MinHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *MinHeap) Push(x interface{}) { *h = append(*h, x.(*OrderEntry)) }
func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// orderbook
type Orderbook struct {
	Market       string
	Asks         *MinHeap
	Bids         *MaxHeap
	CurrentPrice float64
}

func (ob *Orderbook) Buy(price float64, quantity float64) {
	if ob.Bids.Len() == 0 {heap.Push(ob.Bids, &OrderEntry{Price : price, Quantity: quantity})}

	bestAsk := (*ob.Asks)[0]
	bestBid := (*ob.Bids)[0]

	if price >= bestAsk.Price {
		tempquantity := quantity
	
		for _,value := range *ob.Asks {
			if(tempquantity >= value.Quantity){
				tempquantity -= value.Quantity
				// heap.Pop(ob.Asks);
				ob.Asks.Pop()
			}
			if (tempquantity < value.Quantity && price >=value.Price){
				value.Quantity -= tempquantity;
				ob.CurrentPrice = value.Price;
				tempquantity = 0;
				break;
			}

		}

		if (tempquantity > 0 ) {
			// order := OrderEntry{Price: price, Quantity: quantity}
			// ob.Bids.Push(order);
			ob.Bids.Push([]interface{}{price, quantity})

		}
	}
	if (price < bestAsk.Price){
		flag := false;
		if(price > bestBid.Price){ ob.Bids.Push([]interface{}{price, quantity});}
		if(price <= bestBid.Price){
			for _,value := range *ob.Bids {
				if(price == value.Price){
					flag = true;
					value.Quantity += quantity;
					break;

				}
			}
			if(!flag){ob.Bids.Push([]interface{}{price,quantity});}

			if ob.CurrentPrice == 0 {
				if bestAsk != nil  {
					ob.CurrentPrice = bestAsk.Price;
				}
			}
			
		}


		
	}


}
func (ob *Orderbook) Sell(price float64, quantity float64) {
	if ob.Asks.Len() == 0 {heap.Push(ob.Asks, &OrderEntry{Price: price, Quantity: quantity})}

	bestAsk := (*ob.Asks)[0]
	bestBid := (*ob.Bids)[0]

	if price <= bestBid.Price {
		tempquantity := quantity

		for _, value := range *ob.Bids {
			if tempquantity >= value.Quantity {
				tempquantity -= value.Quantity
				ob.Bids.Pop()
			}
			if tempquantity < value.Quantity && price <= value.Price {
				value.Quantity -= tempquantity
				ob.CurrentPrice = value.Price
				tempquantity = 0
				break
			}
		}

		if tempquantity > 0 {
			ob.Asks.Push([]interface{}{price, quantity})
		}
	}
	if price > bestBid.Price {
		flag := false
		if price < bestAsk.Price {
			ob.Asks.Push([]interface{}{price, quantity})
		}
		if price >= bestAsk.Price {
			for _, value := range *ob.Asks {
				if price == value.Price {
					flag = true
					value.Quantity += quantity
					break
				}
			}
			if !flag {
				ob.Asks.Push([]interface{}{price, quantity})
			}

			if ob.CurrentPrice == 0 {
				if bestBid != nil {
					ob.CurrentPrice = bestBid.Price
				}
			}
		}
	}
}


func (ob *Orderbook) GetDepth() ([]*OrderEntry, []*OrderEntry){
	bids := make([]*OrderEntry, len(*ob.Bids))
	copy(bids, *ob.Bids)

	asks := make([]*OrderEntry, len(*ob.Asks))
	copy(asks, *ob.Asks)

	return  bids, asks
}
